x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/extractor/CompositeKeyExtractor.java

```diff
@@ -95,7 +95,9 @@ public Object extract(Bucket bucket) {
             if (object == null) {
                 return object;
             } else if (object instanceof Long) {
-                object = DateUtils.of(((Long) object).longValue(), zoneId);
+                object = DateUtils.asDateTime(((Long) object).longValue(), zoneId);
+            } else if (object instanceof String) { // CAST(<value> AS DATE) is used
```

Comment:
This looks suspicious - why would the key be a string and that would represent only a date and not a datetime?

Reply:
If `zoneId != null` we now we are expecting a DATETIME or DATE.
If it's a Long it's a DATETIME (ES date), If it's a DATE then there is a CAST applied with painless and the object extracted is a String. Do you have some other idea? Maybe I'm missing something.

Reply:
The painless script is unnecessary and complicated namely because a histogram with a granularity of 1 day is going to ignore the time component _anyway_. We can and should identify this pattern `GROUP BY - CAST(DATETIME->DATE)` and simply convert that to a `GROUP BY (DATETIME)`.
I'm not sure if the time will show-up or not (I would expect it to be zero anyway due to 1 day granularity).
Even if we were to use the painless script, since the casting is to a date, it should still return a `ZonedDateTime` and not a `String` however as I mention, we can optimize the script away.

I'm fine with addressing this in a different PR.